### ISSUE TITLE
MissingObjectException in GitTreeEntryIterator

### ISSUE THREAD

**Issue by smacker** [src-d/engine#332]
## Expected Behavior

GitTreeEntryIterator works similar to CommitIterator. When it meets MissingObjectException, it logs warning and continues working.

## Current Behavior

A task failing with exception like this:

```
org.eclipse.jgit.errors.MissingObjectException: Missing tree 68aafcebbf6f67a7648b2d5b3bac8c2db70a0867
	at org.eclipse.jgit.internal.storage.file.WindowCursor.open(WindowCursor.java:164)
	at org.eclipse.jgit.treewalk.CanonicalTreeParser.reset(CanonicalTreeParser.java:214)
	at org.eclipse.jgit.treewalk.CanonicalTreeParser.createSubtreeIterator0(CanonicalTreeParser.java:248)
	at org.eclipse.jgit.treewalk.CanonicalTreeParser.createSubtreeIterator(CanonicalTreeParser.java:226)
	at org.eclipse.jgit.treewalk.CanonicalTreeParser.createSubtreeIterator(CanonicalTreeParser.java:70)
	at org.eclipse.jgit.treewalk.TreeWalk.enterSubtree(TreeWalk.java:1271)
	at tech.sourced.engine.iterator.GitTreeEntryIterator$$anonfun$tech$sourced$engine$iterator$GitTreeEntryIterator$$getTreeEntries$3.apply(GitTreeEntryIterator.scala:152)
	at tech.sourced.engine.iterator.GitTreeEntryIterator$$anonfun$tech$sourced$engine$iterator$GitTreeEntryIterator$$getTreeEntries$3.apply(GitTreeEntryIterator.scala:150)
	at scala.collection.immutable.Stream$$anonfun$map$1.apply(Stream.scala:418)
	at scala.collection.immutable.Stream$$anonfun$map$1.apply(Stream.scala:418)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1233)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1223)
	at scala.collection.immutable.Stream$$anonfun$filteredTail$1.apply(Stream.scala:1299)
	at scala.collection.immutable.Stream$$anonfun$filteredTail$1.apply(Stream.scala:1299)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1233)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1223)
	at scala.collection.immutable.Stream$$anonfun$map$1.apply(Stream.scala:418)
	at scala.collection.immutable.Stream$$anonfun$map$1.apply(Stream.scala:418)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1233)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1223)
	at scala.collection.immutable.StreamIterator$$anonfun$next$1.apply(Stream.scala:1120)
	at scala.collection.immutable.StreamIterator$$anonfun$next$1.apply(Stream.scala:1120)
	at scala.collection.immutable.StreamIterator$LazyCell.v$lzycompute(Stream.scala:1109)
	at scala.collection.immutable.StreamIterator$LazyCell.v(Stream.scala:1109)
	at scala.collection.immutable.StreamIterator.hasNext(Stream.scala:1114)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at tech.sourced.engine.iterator.ChainableIterator.hasNext(ChainableIterator.scala:89)
	at scala.collection.Iterator$class.isEmpty(Iterator.scala:330)
	at tech.sourced.engine.iterator.ChainableIterator.isEmpty(ChainableIterator.scala:16)
	at tech.sourced.engine.iterator.ChainableIterator.hasNext(ChainableIterator.scala:90)
	at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
	at tech.sourced.engine.iterator.CleanupIterator.hasNext(CleanupIterator.scala:23)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:438)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIterator.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anonfun$8$$anon$1.hasNext(WholeStageCodegenExec.scala:395)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:408)
	at org.apache.spark.rdd.RDD$$anonfun$countApprox$1$$anonfun$27.apply(RDD.scala:1180)
	at org.apache.spark.rdd.RDD$$anonfun$countApprox$1$$anonfun$27.apply(RDD.scala:1178)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
	at org.apache.spark.scheduler.Task.run(Task.scala:108)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:335)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

And after mutliple failings stage gets killed.

## Steps to Reproduce (for bugs)

```
spark-shell --packages "tech.sourced:engine:0.5.1" --master="spark://p-spark-master:7077"

scala> import tech.sourced.engine._
import tech.sourced.engine._

scala> val engine = Engine(spark, "hdfs://hdfs-namenode/apps/borges/10k", "siva")
engine: tech.sourced.engine.Engine = tech.sourced.engine.Engine@4086e003

scala> engine.getRepositories.getHEAD.getCommits.getTreeEntries.count()
[Stage 0:=====>                                                 (28 + 96) / 298]18/02/26 14:06:56 WARN TaskSetManager: Lost task 80.0 in stage 0.0 (TID 81, 10.2.101.146, executor 0): org.eclipse.jgit.errors.MissingObjectException: Missing tree 68aafcebbf6f67a7648b2d5b3bac8c2db70a0867
```

## Context

Run gemini

## Your Environment (for bugs)

* Version used: 0.5.1
* Src-d staging cluster

**Comment by ajnavarro** [comment]
Do you have any information about the siva file that is causing the error?

**Comment by smacker** [comment]
Nope. There are 10k of them. And traceback doesn't provide any info about siva file.

**Comment by ajnavarro** [comment]
Then, can you add the complete log instead of only the stacktrace?

**Comment by smacker** [comment]
Sorry I don't understand what log do you need. Anyway here is all output: https://gist.github.com/smacker/ebaf1bf3aed30ca70190e8b96f138a35

it contains only traceback from above + Lost task warnings.
spark doesn't produce anything else.

**Comment by ajnavarro** [comment]
Steps to get the logs:
- Set log level to DEBUG
- Execute the process that is failing
- Get logs of all the workers that are executing the process
- Add one of the logs of one of the workers that contains the error stacktrace of this issue here

**Comment by smacker** [comment]
Thanks. But could you point to any link how to do that?

I tried to set "DEBUG" using `sc.setLogLevel("DEBUG")` as spark recommends.
It produces this output: https://gist.github.com/smacker/e37868781f64e07f735908e875c6e35b

But looks like it isn't what you need.

**Comment by ajnavarro** [comment]
At the end, It appears to be two objects that are missing: 
- 68aafcebbf6f67a7648b2d5b3bac8c2db70a0867
- b49e084d08c1f5dab9437f4d5d91925777209836

With the actual logs, that are not complete, I could say that:
- Can be one or two siva files corrupted
- One or two siva files have some corner cases that we didn't implement

**Comment by ajnavarro** [comment]
> Thanks. But could you point to any link how to do that?
  I tried to set "DEBUG" using sc.setLogLevel("DEBUG") as spark recommends.
  It produces this output: gist.github.com/smacker/e37868781f64e07f735908e875c6e35b
  But looks like it isn't what you need.

I need the logs of each worker. The process is not executed where the spark-shell is,  that is the driver. I need the logs for the place where the faulty code is executed, in this case one of the workers.

**Comment by smacker** [comment]
Please give me information how to collect that logs. I did quick googling and can't figure out it.

Only one related thing I have found:

I see failed tasks in Spark UI. Here is stdout of one of them: http://localhost:8001/spark-ui-proxy/api/v1/namespaces/default/services/p-spark-webui-proxy:80/proxy/proxy:10.2.47.2:8081/logPage/?appId=app-20180226151259-0065&executorId=2&logType=stderr
The tasks: http://localhost:8001/spark-ui-proxy/api/v1/namespaces/default/services/p-spark-webui-proxy:80/proxy/proxy:10.2.36.34:4040/stages/stage?id=0&attempt=0

Links are alive only as long as I don't kill the application as I understand.

If it's not what you need, just give me instructions, I'll run it and collect logs.

**Comment by ajnavarro** [comment]
I went to the [spark documentation](https://spark.apache.org/docs/latest/spark-standalone.html) and found:


> -d DIR, --work-dir DIR | Directory to use for scratch space and job output logs 
  (default: SPARK_HOME/work); only on worker

Then, to know how is configured our cluster, I went to:
> https://github.com/src-d/charts/blob/master/spark/templates/spark-worker-daemonset.yaml#L25

And check that the spark home is `/opt/spark/sbin/`

Then, as the documentation is saying, go to `work/` directory. There we have a lot of subfolders, one per each execution. I opened `work/app-20180226151259-0065/1` and check the files. There we have `stderr` file.
